rust/src/array.rs

```diff
@@ -18,447 +18,562 @@
-                }
-            },
-            &None => for i in 0..self.len {
-                let mut m = self.data.get(i);
-                match n {
-                    None => n = Some(*m),
-                    Some(nn) => if *m < nn {
-                        n = Some(*m)
-                    },
-                }
-            },
+/// Constructs a `ListArray` from an array data reference.
+impl From<ArrayDataRef> for ListArray {
+    fn from(data: ArrayDataRef) -> Self {
+        debug_assert!(data.buffers().len() == 1);
+        debug_assert!(data.child_data().len() == 1);
+        let values = make_array(data.child_data()[0].clone());
+        let value_offsets = data.buffers()[0].raw_data() as *const i32;
+        unsafe {
+            debug_assert!(*value_offsets.offset(0) == 0);
+            debug_assert!(*value_offsets.offset(data.length() as isize)
+                          == values.data().length() as i32);
+        }
+        Self {
+            data: data.clone(),
+            values: values,
+            value_offsets: RawPtrBox::new(value_offsets),
         }
-        n
+    }
+}
+
+impl Array for ListArray {
+    fn as_any(&self) -> &Any {
+        self
     }
 
-    /// Determine the maximum value in the array
-    pub fn max(&self) -> Option<T> {
-        let mut n: Option<T> = None;
-        match &self.validity_bitmap {
-            &Some(ref bitmap) => for i in 0..self.len {
-                if bitmap.is_set(i) {
-                    let mut m = self.data.get(i);
-                    match n {
-                        None => n = Some(*m),
-                        Some(nn) => if *m > nn {
-                            n = Some(*m)
-                        },
-                    }
-                }
-            },
-            &None => for i in 0..self.len {
-                let mut m = self.data.get(i);
-                match n {
-                    None => n = Some(*m),
-                    Some(nn) => if *m > nn {
-                        n = Some(*m)
-                    },
-                }
-            },
-        }
-        n
+    fn data(&self) -> ArrayDataRef {
+        self.data.clone()
+    }
+
+    fn data_ref(&self) -> &ArrayDataRef {
+        &self.data
     }
 }
 
-/// Implement the Add operation for types that support Add
-impl<T> PrimitiveArray<T>
-where
-    T: ArrowPrimitiveType + Add<Output = T>,
-{
-    pub fn add(&self, other: &PrimitiveArray<T>) -> PrimitiveArray<T> {
-        let mut builder: Builder<T> = Builder::new();
-        for i in 0..self.len {
-            let x = *self.data.get(i) + *other.data.get(i);
-            builder.push(x);
+/// A special type of `ListArray` whose elements are binaries.
+pub struct BinaryArray {
+    data: ArrayDataRef,
+    value_offsets: RawPtrBox<i32>,
+    value_data: RawPtrBox<u8>,
+}
+
+impl BinaryArray {
+    /// Returns the element at index `i` as a byte slice.
+    pub fn get_value(&self, mut i: i64) -> &[u8] {
+        i += self.data.offset();
```

Comment:
uh, that may overflow in production builds (overflows in Rust are only checked in debug builds) and lead to a security issue some lines below, because you read from a out-of-bounds position. Please add some checks here that this doesn't happen:

1. check that the integer addition doesn't overflow ( https://doc.rust-lang.org/std/primitive.usize.html#method.checked_add )
2. check that the initial `i` is actually in-range.

Reply:
Good point! fixed.